fedbiomed/common/config.py

```diff
@@ -0,0 +1,210 @@
+
+        Raises verision compatibility error
+        """
+
+        self._cfg.read(self.path)
+
+        # Validate config version
+        raise_for_version_compatibility(
+            self._cfg["default"]["version"],
+            self.CONFIG_VERSION,
+            f"Configuration file {self.path}: found version %s expected version %s")
+
+        return True
+
+    def get(self, section, key) -> str:
+        """Returns value for given ket and section"""
+
+        return self._cfg.get(section, key)
+
+    def sections(self) -> list:
+        """Returns sections of the config"""
+
+        return self._cfg.sections()
+
+    def write(self):
+        """Writes config file"""
+
+        try:
+            with open(self.path, 'w') as f:
+                self._cfg.write(f)
+        except configparser.Error:
+            raise IOError(ErrorNumbers.FB600.value + ": cannot save config file: " + self.path)
+
+
+    def generate(self, force: bool = False) -> bool:
+        """"Generate configuration file
+
+        Args:
+        force: Overwrites existing configration file
+        """
+
+        # Check if configuration is already existing
+        if self.is_config_existing() and not force:
+            return self.read()
+
+        component_id = f"{self.COMPONENT_TYPE}_{uuid.uuid4()}"
+
+        self._cfg['default'] = {
+            'id': component_id,
+            'component': self.COMPONENT_TYPE,
+            'version': str(self.CONFIG_VERSION)
+        }
+
+        # DB PATH RELATIVE
+        db_path  = os.path.join(self.root, VAR_FOLDER_NAME, f"{DB_PREFIX}{component_id}.json")
+        self._cfg['default']['db'] = os.path.relpath(db_path, os.path.join(self.root, CONFIG_FOLDER_NAME))
+
+        ip, port = retrieve_ip_and_port(self.root)
+        allow_default_biprimes = os.getenv('ALLOW_DEFAULT_BIPRIMES', True)
+
+        # Generate self-signed certificates
+        key_file, pem_file = generate_certificate(
+            root=self.root, 
+            component_id=component_id, 
+            prefix=MPSPDZ_certificate_prefix)
+
+        self._cfg['mpspdz'] = {
+            'private_key': os.path.relpath(key_file, os.path.join(self.root, 'etc')),
+            'public_key': os.path.relpath(pem_file, os.path.join(self.root, 'etc')),
+            'mpspdz_ip': ip,
+            'mpspdz_port': port,
+            'allow_default_biprimes': allow_default_biprimes,
+            'default_biprimes_dir': os.path.relpath(
+                os.path.join(self.root, 'envs', 'common', 'default_biprimes'),
+                os.path.join(self.root, 'etc')
+            )
+        }
+
+        # Calls child class add_parameterss
+        self.add_parameters()
+
+        # Write configuration file
+        return self.write()
+
+    @abstractmethod
+    def add_parameters(self):
+        """"Component specific argument creation"""
+
+
+class NodeConfig(Config):
+
+    DEFAULT_CONFIG_FILE_NAME: str = 'config_node.ini'
```

Comment:
Wouldn't `DEFAULT_CONFIG_FILE_NAME` `COMPONENT_TYPE` `CONFIG_VERSION` gain to be private variables ?
They are only used within the class (and parent class).

Same in `ResearcherConfig`.

Reply:
I consider those as constant attributes of the config classes.  I think making it private is very little detail  but if you say that it is not clear to understand or confusing I can update make it private.

Reply:
Thank you Sergen,
Yes it would be nice to make them private to clarify they are constants.

Reply:
Sorry you right it is good to make it private to clearly state for developers that is not meant to change.

Reply:
I pushed a new commit for this.